File: 07-trees/00100-Same_Tree/00100-same_tree.go
// 100: Same Tree
// https://leetcode.com/problems/same-tree/

package main

import "fmt"

type TreeNode struct {
    val any
    left *TreeNode
    right *TreeNode
}

type Tree struct {
    root *TreeNode
}

func creator(values []any, root **TreeNode, i, n int) *TreeNode {
    if n==0 {return nil}
    if i<n {
        temp := &TreeNode{values[i], nil, nil}
        *root = temp
        (*root).left = creator(values, &((*root).left), 2*i+1, n);
        (*root).right = creator(values, &((*root).right), 2*i+2, n);
    }
    return *root;
}

func createTree(root **TreeNode, inputs []any) {
    creator(inputs, root, 0, len(inputs))
}

func showTree(root *TreeNode) {
    var q []*TreeNode;
    var result [][]interface{}
    var c []interface{}
    if root==nil { fmt.Println("Empty !"); return; }
    q = append(q, root)
    q = append(q, nil)
    for len(q)!=0 {
        t := q[0]
        q = q[1:]
        if t==nil {
            result = append(result, c)
            c = make([]interface{}, 0)
            if len(q) > 0 {q = append(q, nil)}
        } else {
            c = append(c, t.val)
            if t.left!=nil {q = append(q, t.left)}
            if t.right!=nil {q = append(q, t.right)}
        }
    }

    fmt.Print("["); for _, x := range result {
        fmt.Print("["); for _, y := range x {
            if y==nil { fmt.Print("NULL,"); continue; }
            fmt.Print(y,",")
        }; fmt.Print("\b],")
    }; fmt.Println("\b]")
}

// SOLUTION
func isSameTree(p, q *TreeNode) bool {
    if p == nil && q == nil {return true}
    if p == nil || q == nil {return false}
    if p.val != q.val {return false}
    return isSameTree(p.left, q.left) && isSameTree(p.right, q.right)
}



func main() {
    p, q := Tree{}, Tree{}

    // INPUT
    pn, qn := []any{1,2,3}, []any{1,2,3}
    createTree(&p.root, pn)
    createTree(&q.root, qn)

    // OUTPUT
    result := isSameTree(p.root, q.root)
    fmt.Println(result)
}

File: 07-trees/00100-Same_Tree/00100-same_tree.py
# 100: Same Tree
# https://leetcode.com/problems/same-tree/



class TreeNode:
    def __init__(self, x: int):
        self.val = x
        self.left = None
        self.right = None

class Tree:
    def __init__(self) -> None:
        self.root = None

    def creator(self, values: list[int], root: TreeNode, i: int, n: int) -> TreeNode:
        if n==0 : return None
        if i<n:
            temp = TreeNode(values[i])
            root = temp
            root.left = self.creator(values, root.left, 2*i+1, n)
            root.right = self.creator(values, root.right, 2*i+2, n)
        return root

    def createTree(self, inputs: list[int]) -> None:
        self.root = self.creator(inputs, self.root, 0, len(inputs))
    
    def showTree(self, root: TreeNode) -> None:
        q = []
        result = [[]]
        c = []
        if root==None : return result
        q.append(root)
        q.append(None)
        while len(q)!=0:
            t = q.pop(0)
            if t==None:
                result.append(c)
                c = []
                if len(q) > 0 : q.append(None)
            else:
                c.append(t.val)
                if t.left!=None : q.append(t.left)
                if t.right!=None : q.append(t.right)

        print("[", end="")
        for x in result:
            if len(x)==0 : continue
            print("[", end="")
            for y in x:
                if y==None: 
                    print("NULL,", end="")
                    continue
                print(y, end=",")
            print("\b]", end=",")
        print("\b]")


class Solution:
    # SOLUTION
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        if p == None and q == None : return True
        if p == None or q == None : return False
        if p.val != q.val : return False
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)



if __name__ == "__main__":
    o = Solution()
    p, q = Tree(), Tree()

    # INPUT
    pn, qn = [1,2,3], [1,2,3]
    p.createTree(pn)
    q.createTree(qn)

    # OUTPUT
    result = o.isSameTree(p.root, q.root)
    print(result)
